content/articles/2017-11-15-withr-2.1.0.Rmd

```diff
@@ -0,0 +1,234 @@
+less error prone. It also has no non-base dependencies so is easy to include in
+packages.
+
+Install the latest version with:
+```{r, eval = FALSE}
+install.packages("withr")
+```
+
+# Why withr?
+
+Many functions in R modify global state in some fashion. Some common examples
+are `par()` for graphics parameters, `dir()` to change the current directory
+and `options()` to set a global option. Using these global options is useful
+when using R interactively, because you can set them early in your
+experimentation and they will remain set for the duration of the session.
+However this makes programming with these settings difficult, because in a
+pure function you should always reset the previous state when the function
+completes if you change a global setting.
+
+One common idiom for dealing with this problem is to save the current state,
+make your change, then reset the previous state.
+
+```{r}
+par("col" = "black")
+my_plot <- function(new) {
+  old <- par(col = "red", pch = 19)
+  plot(mtcars$hp, mtcars$wt)
+  par(old)
+}
+my_plot()
+par("col")
+```
+
+However one issue when there is an error before you are able to reset the
+options.
+
+```{r, error = TRUE}
+par("col" = "black")
+my_plot <- function(new) {
+  old <- par(col = "red", pch = 19)
+  plot(mtcars$hpp, mtcars$wt)
+  par(old)
+}
+my_plot()
+par("col")
+```
+
+Using the base function `on.exit()` is the robust solution to this problem,
+which will run the resetting code when the block is exited, regardless if it
+exits normally or with an error.
+
+```{r, error = TRUE}
+par("col" = "black")
+my_plot <- function(new) {
+  old <- par(col = "red", pch = 19)
+  on.exit(par(old))
+  plot(mtcars$hpp, mtcars$wt)
+}
+my_plot()
+par("col")
+```
+
+However this solution is somewhat cumbersome to work with. You
+need to remember to use an `on.exit()` call after each stateful call. In
+addition by default each `on.exit()` action will overwrite any previous
+`on.exit()` action in the same block (using `add = TRUE` prevents this). It is also
+not possible to have this cleanup performed before the block has finished.
+
+[withr](http://withr.r-lib.org) is a solution to these issues. It defines a
+[large set of
+functions](http://withr.r-lib.org/#withr---run-code-with-modified-state) for
+dealing with global settings in R, such as `with_par()`. These functions set one of
+the global settings for the duration of a block of code, then automatically
+reset it after the block is completed.
+
+```{r}
+par("col" = "black")
+my_plot <- function(new) {
+  with_par(list(col = "red", pch = 19),
+    plot(mtcars$hp, mtcars$wt)
+  )
+  par("col")
+}
+my_plot()
+par("col")
+```
+
+In addition to the `with_*` functions there are `local_*` variants which simply
+last until the end of the block they are included in. These work similar to
+`on.exit()`, but you can set the options in one call rather than two.
+
+```{r}
+par("col" = "black")
+my_plot <- function(new) {
+  local_par(list(col = "red", pch = 19))
```

Comment:
I kind of wish these took `...` instead of requiring a list

Reply:
Unfortunately supporting this makes the implementations much more complicated, you need to then check which arguments are named, make sure there is only one etc. There also could be name clashes with regular arguments for things like `with_options()`.

While I agree the syntax is slightly more cumbersome than otherwise, I think the added consistency and simplicity makes it worthwhile.

Reply:
Yeah, I understand, I just still wish they took `...` 😉